### DISCUSSION TITLE
Separate Learning Rates for PPO

### DISCUSSION THREAD

**Discussion by hiraz01** [Toni-SM/skrl#368] [Q&A]
Is it possible to use two separate optimizers and learning rates for the policy and value optimizers without directly changing the current PPO implementation? I looked through the source and it appears that I could change it directly, but I am not sure how best way to go about doing this. I am trying to replicate results from DeepMimic environment where they use separate LRs in tensorflow 1. Goal is to have additional flexibility and use IsaacLab with skrl.

**Comment by Toni-SM** [comment]
Hi @hiraz01 

Currently, it is necessary to modify the source code to support different learning rate for the involved optimizers.
But your comment gave me the idea to support, as a learning rate value, a tuple too (where each element is the learning rate of each optimizer). I hope to have some time to develop it :sweat_smile:

**Reply by hiraz01** [reply]
I think that could be useful. From what I can gather, none of the major libraries out there provide this by default (e.g. SB3 and RLLib). I found a PyTorch implementation that addresses this but does not include all the optimizations. For now, I added to the configuration and changed a couple lines at the optimizer initialization : 

```python
self._split_lr = self.cfg["split_lr"]
self._policy_learning_rate = self.cfg["policy_learning_rate"]
self._value_learning_rate = self.cfg["value_learning_rate"]
```

In the __init__ section I have,

```python
elif self._split_lr == False:
     self.optimizer = torch.optim.Adam(itertools.chain(self.policy.parameters(), self.value.parameters()), lr=self._learning_rate)
else:
     self.optimizer = torch.optim.Adam([{'params': self.policy.parameters(), 'lr': self._policy_learning_rate},
                                                               {'params': self.value.parameters(), 'lr': self._value_learning_rate}])
```

I think this is all that needed to change to accomplish a split learning rate using the same optimizer. Correct me if I am wrong, but we would also still need to change the learning rate scheduler to perform any progress-based learning rate changes. It would be nice to verify this is correct, and in the future have a setup where we could also potentially use two separate optimizers for full flexibility. I am happy to help with testing, developing, or in any other way.

**Reply by Toni-SM** [reply]
~I am not sure that specifying `'lr'` in the list of dictionaries will work.~
I can not find a reference to it in PyTorch's [Adam docs](https://docs.pytorch.org/docs/stable/generated/torch.optim.Adam.html#torch.optim.Adam).

---

In any case, you can separate the optimizers and LR schedulers as done for the DDPG implementation:

https://github.com/Toni-SM/skrl/blob/90adbbc1593ebb7ea5f98f39bd22f7e88d1198b2/skrl/agents/torch/ddpg/ddpg.py#L174-L185

**Reply by hiraz01** [reply]
Thank you very much for your response. I am not sure either, but I did find this library where they use this method for implementation.

https://github.com/nikhilbarhate99/PPO-PyTorch/blob/master/PPO.py

In lines 139 to 142 of the PPO.py:
```python
        self.optimizer = torch.optim.Adam([
                        {'params': self.policy.actor.parameters(), 'lr': lr_actor},
                        {'params': self.policy.critic.parameters(), 'lr': lr_critic}
                    ])
```
Then in the update step they make a single call to zero_grad and step as usual. I was hoping to do more checking and drilling down to see if this does, in fact work, as written. I will have a look at the Adam link you provided and continue to research this. Thanks again! The library and your responses are much appreciated.

**Reply by Toni-SM** [reply]
Ohhh, I see...

Looking deeper into the PyTorch's Adam base class ([Optimizer](https://docs.pytorch.org/docs/stable/optim.html#torch.optim.Optimizer))... the learning rate is updated to `param_group` in https://github.com/pytorch/pytorch/blob/f638854e1da6b33e78dcc9f3e28c98c4cdce4e86/torch/optim/optimizer.py#L1124-L1130. So, yes, it is possible to specify the learning rate using dict :eyes: (something new to learn today!)

Although I must say that the implementation is quite obscure and the PyTorch documentation does not clarify the details very well. [Optimizer](https://docs.pytorch.org/docs/stable/optim.html#torch.optim.Optimizer) docs:
> params (iterable) – an iterable of [torch.Tensor](https://docs.pytorch.org/docs/stable/tensors.html#torch.Tensor) s or [dict](https://docs.python.org/3/library/stdtypes.html#dict) s. Specifies what Tensors should be optimized.
> defaults ([dict](https://docs.python.org/3/library/stdtypes.html#dict)[[str](https://docs.python.org/3/library/stdtypes.html#str), [Any](https://docs.python.org/3/library/typing.html#typing.Any)]) – (dict): a dict containing default values of optimization options (used when a parameter group doesn’t specify them).

**Reply by hiraz01** [reply]
That is good to hear. Sometimes a combination of searching engines and library code and even PyTorch documentation does not always lead me to a very confident conclusion. I think that my first next step will be to use the single Adam optimizer, in combination with two learning rates defined using the dict, as we have discussed, and with a single learning rate schedule to adjust both rates. Thanks again.
